File: java_projects/glue_code/commons-fileupload/src/main/java/org/apache/commons/fileupload/util/LimitedInputStream.java
package org.apache.commons.fileupload.util;
import org.apache.commons.fileupload.ExceptionHandler;

import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;
import org.apache.commons.fileupload.ContextInitializer;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

public abstract class LimitedInputStream extends FilterInputStream implements Closeable {
  private static Value clz =
      ContextInitializer.getPythonClass("/utiLimitedInputStream.py", "LimitedInputStream");
  private Value obj;

  // public LimitedInputStream(Value obj) {
  //   this.obj = obj;
  // }

  public Value getPythonObject() {
    return obj;
  }

  public void close() throws IOException {
    try {
      //
      // closed = true;
      // super.close();
      //

      obj.invokeMember("close");
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "LimitedInputStream.close");
    }
  }

  public boolean isClosed() throws IOException {
    try {
      //
      // return closed;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("isClosed").as(boolean.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "LimitedInputStream.isClosed");
    }
  }

  public int read1(byte[] b, int off, int len) throws IOException {
    try {
      //
      // int res = super.read(b, off, len);
      // if (res > 0) {
      // count += res;
      // checkLimit();
      // }
      // return res;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("read1", b, off, len).as(int.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "LimitedInputStream.read1");
    }
  }

  public int read0() throws IOException {
    try {
      //
      // int res = super.read();
      // if (res != -1) {
      // count++;
      // checkLimit();
      // }
      // return res;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("read0").as(int.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "LimitedInputStream.read0");
    }
  }

  public LimitedInputStream(InputStream inputStream, long pSizeMax) {
    super(inputStream);
    // sizeMax = pSizeMax;
    //

    this.obj = clz.invokeMember("__init__", inputStream, pSizeMax);
  }

  private void checkLimit() throws IOException {
    try {
      //
      // if (count > sizeMax) {
      // raiseError(sizeMax, count);
      // }
      //

      obj.invokeMember("checkLimit");
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "LimitedInputStream.checkLimit");
    }
  }

  protected abstract void raiseError(long pSizeMax, long pCount) throws IOException;
}

File: data/verified_projects/commons-fileupload/src/main/org/apache/commons/fileupload/util/LimitedInputStream.py
import pytest

# Imports Begin
from src.main.org.apache.commons.fileupload.util.Closeable import *
import os
import typing
from typing import *
import io
from abc import ABC

# Imports End


class LimitedInputStream(Closeable, FilterInputStream, ABC):

    # Class Fields Begin
    __sizeMax: int = None
    __count: int = None
    __closed: bool = None
    # Class Fields End

    # Class Methods Begin
    def close(self) -> None:

        self.__closed = True
        super().close()

    def isClosed(self) -> bool:

        return self.__closed

    def read1(self, b: typing.List[int], off: int, len: int) -> int:

        res = super().read(b, off, len)
        if res > 0:
            self.__count += res
            self.__checkLimit()
        return res

    def read0(self) -> int:

        res = super().read()
        if res != -1:
            self.__count += 1
            self.__checkLimit()
        return res

    def __init__(
        self,
        inputStream: typing.Union[io.BytesIO, io.StringIO, io.BufferedReader],
        pSizeMax: int,
    ) -> None:

        super().__init__(inputStream)
        self.__sizeMax = pSizeMax

    def __checkLimit(self) -> None:

        if self.__count > self.__sizeMax:
            self._raiseError(self.__sizeMax, self.__count)

    def _raiseError(self, pSizeMax: int, pCount: int) -> None:

        raise IOError(f"Maximum size exceeded: {pSizeMax} bytes, {pCount} items")

    # Class Methods End
